pkg/gofr/subscriber.go

```diff
@@ -24,38 +24,57 @@ func newSubscriptionManager(c *container.Container) SubscriptionManager {
 	}
 }
 
-func (s *SubscriptionManager) startSubscriber(topic string, handler SubscribeFunc) {
+func (s *SubscriptionManager) startSubscriber(ctx context.Context, topic string, handler SubscribeFunc) {
 	// continuously subscribe in an infinite loop
 	for {
-		msg, err := s.container.GetSubscriber().Subscribe(context.Background(), topic)
-		if msg == nil {
-			continue
-		}
-
-		if errors.Is(err, kafka.ErrConsumerGroupNotProvided) {
-			s.container.Logger.Errorf("cannot subscribe as consumer_id is not provided in configs")
+		select {
+		case <-ctx.Done():
+			s.container.Logger.Infof("shutting down subscriber for topic %s", topic)
 			return
-		} else if err != nil {
-			s.container.Logger.Errorf("error while reading from topic %v, err: %v", topic, err.Error())
-			continue
+		default:
+			s.handleSubscription(ctx, topic, handler)
 		}
+	}
+}
 
-		ctx := newContext(nil, msg, s.container)
-		err = func(ctx *Context) error {
-			// TODO : Move panic recovery at central location which will manage for all the different cases.
-			defer func() {
-				panicRecovery(recover(), ctx.Logger)
-			}()
-
-			return handler(ctx)
-		}(ctx)
-
-		// commit the message if the subscription function does not return error
-		if err == nil {
-			msg.Commit()
-		} else {
-			s.container.Logger.Errorf("error in handler for topic %s: %v", topic, err)
-		}
+func (s *SubscriptionManager) handleSubscription(parentCtx context.Context, topic string, handler SubscribeFunc) {
+	ctx, done := context.WithCancel(parentCtx)
+	defer done()
+
+	msg, err := s.container.GetSubscriber().Subscribe(ctx, topic)
+
+	if errors.Is(err, kafka.ErrConsumerGroupNotProvided) {
+		s.container.Logger.Errorf("cannot subscribe as consumer_id is not provided in configs")
```

Comment:
This one doesn't need an Errorf


```suggestion
		s.container.Logger.Error("cannot subscribe as consumer_id is not provided in configs")
```